File: Leetcode/Dart/66.dart
class Solution {
  List<int> plusOne(List<int> digits) {
    digits.insert(0, 0); // 避免進位
    int n = digits.length;
    int last = n - 1; // 最後一個數字的 index
    digits[last] += 1;
    while (digits[last] >= 10) {
      digits[last] -= 10;
      last -= 1;
      digits[last] += 1;
    }
    return digits[0] == 0 ? digits.sublist(1) : digits;
  }
}

void main(){
  Solution solution = new Solution();
  print(solution.plusOne([1,2,3])); // [1,2,4]
  print(solution.plusOne([4,3,2,1])); // [4,3,2,2]
  print(solution.plusOne([0])); // [1]
  print(solution.plusOne([9])); // [1,0]
}

File: Leetcode/cpp/easy/66.cpp
/*
 * @lc app=leetcode.cn id=66 lang=cpp
 * @lcpr version=30204
 *
 * [66] 加一
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        digits[n - 1] += 1;
        int i = n - 1;
        while (i >= 0) {
            if (digits[i] < 10) {
                break;
            }
            digits[i] -= 10;
            if (i == 0) {
                digits.insert(digits.begin(), 1);
                break;
            }
            digits[i - 1] += 1;
            i -= 1;
        }
        return digits;
    }
};
// @lc code=end



/*
// @lcpr case=start
// [1,2,3]\n
// @lcpr case=end

// @lcpr case=start
// [4,3,2,1]\n
// @lcpr case=end

// @lcpr case=start
// [9]\n
// @lcpr case=end

 */

